processor/src/main/java/org/mapstruct/ap/internal/model/StreamMappingMethod.java

```diff
@@ -0,0 +1,404 @@
+
+        private Assignment forgeMapping(SourceRHS sourceRHS, Type sourceType, Type targetType) {
+
+            ForgedMethodHistory forgedMethodHistory = null;
+            if ( method instanceof ForgedMethod ) {
+                forgedMethodHistory = ( (ForgedMethod) method ).getHistory();
+            }
+            String name = getName( sourceType, targetType );
+            forgedMethod = new ForgedMethod(
+                name,
+                sourceType,
+                targetType,
+                method.getMapperConfiguration(),
+                method.getExecutable(),
+                forgedMethodHistory
+            );
+
+            Assignment assignment = new MethodReference(
+                forgedMethod,
+                null,
+                ParameterBinding.fromParameters( forgedMethod.getParameters() )
+            );
+
+            assignment.setAssignment( sourceRHS );
+
+            return assignment;
+        }
+
+        private String getName(Type sourceType, Type targetType) {
+            String fromName = getName( sourceType );
+            String toName = getName( targetType );
+            return Strings.decapitalize( fromName + "To" + toName );
+        }
+
+        private String getName(Type type) {
+            StringBuilder builder = new StringBuilder();
+            for ( Type typeParam : type.getTypeParameters() ) {
+                builder.append( typeParam.getIdentification() );
+            }
+            builder.append( type.getIdentification() );
+            return builder.toString();
+        }
+    }
+
+    private StreamMappingMethod(Method method, Assignment parameterAssignment, MethodReference factoryMethod,
+                                boolean mapNullToDefault, String loopVariableName,
+                                List<LifecycleCallbackMethodReference> beforeMappingReferences,
+                                List<LifecycleCallbackMethodReference> afterMappingReferences,
+                                SelectionParameters selectionParameters, Set<Type> helperImports,
+                                ForgedMethod forgedMethod) {
+        super( method, beforeMappingReferences, afterMappingReferences,
+            forgedMethod == null ? new ArrayList<ForgedMethod>() : Collections.singletonList( forgedMethod )
+        );
+        this.elementAssignment = parameterAssignment;
+        this.factoryMethod = factoryMethod;
+        this.overridden = method.overridesMethod();
+        this.mapNullToDefault = mapNullToDefault;
+        this.loopVariableName = loopVariableName;
+        this.selectionParameters = selectionParameters;
+        this.helperImports = helperImports;
+    }
+
+    public Parameter getSourceParameter() {
+        for ( Parameter parameter : getParameters() ) {
+            if ( !parameter.isMappingTarget() ) {
+                return parameter;
+            }
+        }
+
+        throw new IllegalStateException( "Method " + this + " has no source parameter." );
```

Comment:
Is this a "regular" error situation? If so, we should raise a proper compiler error instead.